File: sdk/v1.0.2/javascript/photos.js
export class Photos {
  constructor(ctx) { this._ctx = ctx; }

  async addPhotoToProject({ photo, tableName, sourceResolutionX, sourceResolutionY, dateTaken, latitude, longitude, filePath, title, clientSideId }) {
    this._ctx.requireTokens();
    const path = `/api/v1/photos/`;
    const { data } = await this._ctx.client.request(path, 'POST', this._ctx.accessToken, this._ctx.refreshToken, JSON.stringify({ photo, table_name: tableName, source_resolution_x: sourceResolutionX, source_resolution_y: sourceResolutionY, date_taken: dateTaken, latitude, longitude, file_path: filePath, title, client_side_id: clientSideId }));
    return data;
  }

  async getPhotoFromProject(tableName, photoId, { keywordListId } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/${encodeURIComponent(photoId)}`;
    const query = new URLSearchParams();
    if (keywordListId !== undefined) query.set('keyword_list_id', keywordListId);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getPhotoFaceDetailsFromProject(tableName, photoId) {
    this._ctx.requireTokens();
    const path = `/api/v1/photos/face_details/${encodeURIComponent(tableName)}/${encodeURIComponent(photoId)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getAllProjectPhotoIds(tableName, { ascOrDesc, lastId, limit } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    if (lastId !== undefined) query.set('last_id', lastId);
    if (limit !== undefined) query.set('limit', limit);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getTopMiddleBottomProjectPhotoIds(tableName, { ascOrDesc } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos_top/${encodeURIComponent(tableName)}`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getTopMiddleBottomProjectPhotosByTableNameSortedByDateRankedKeysetPaginated(tableName, { ascOrDesc, lastId, limit } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos_ranked/${encodeURIComponent(tableName)}/`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    if (lastId !== undefined) query.set('last_id', lastId);
    if (limit !== undefined) query.set('limit', limit);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getProjectMonthYearsWithPhotos(tableName) {
    this._ctx.requireTokens();
    const path = `/api/v1/photo_month_years/${encodeURIComponent(tableName)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getProjectPhotoIdsByMonth(tableName, month, year, { ascOrDesc } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/month/${encodeURIComponent(month)}/year/${encodeURIComponent(year)}`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getProjectPhotoIdsByMonthKeysetPaginated(tableName, month, year, { ascOrDesc, lastId, limit } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/month/${encodeURIComponent(month)}/year/${encodeURIComponent(year)}/`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    if (lastId !== undefined) query.set('last_id', lastId);
    if (limit !== undefined) query.set('limit', limit);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getTopProjectPhotosByTableNameByMonth(tableName, month, year, { ascOrDesc } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/month/${encodeURIComponent(month)}/year/${encodeURIComponent(year)}/top`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getTopMiddleBottomProjectPhotosByTableNameByMonthSortedByDateRankedKeysetPaginated(tableName, month, year, { ascOrDesc, lastId, limit } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/month/${encodeURIComponent(month)}/year/${encodeURIComponent(year)}/ranked/`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    if (lastId !== undefined) query.set('last_id', lastId);
    if (limit !== undefined) query.set('limit', limit);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getProjectPhotoIdsNoDateTakenKeysetPaginated(tableName, { ascOrDesc, lastId, limit } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/date_taken/none/`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    if (lastId !== undefined) query.set('last_id', lastId);
    if (limit !== undefined) query.set('limit', limit);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getTopProjectPhotosByTableNameNoDateTakenRanked(tableName, { ascOrDesc } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/date_taken/none/ranked`;
    const query = new URLSearchParams();
    if (ascOrDesc !== undefined) query.set('asc_or_desc', ascOrDesc);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getProjectThumbnail(tableName) {
    this._ctx.requireTokens();
    const path = `/api/v1/photos_project/${encodeURIComponent(tableName)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async updatePhotoInProject({ tableName, photoId, photoData } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos_update`;
    const query = new URLSearchParams();
    if (tableName !== undefined) query.set('table_name', tableName);
    if (photoId !== undefined) query.set('photo_id', photoId);
    if (photoData !== undefined) query.set('photo_data', photoData);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'PUT', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async updatePhotoRanking(tableName, photoId, newCategory) {
    this._ctx.requireTokens();
    const path = `/api/v1/photos_update/${encodeURIComponent(tableName)}/id/${encodeURIComponent(photoId)}/rank/${encodeURIComponent(newCategory)}`;
    const { data } = await this._ctx.client.request(path, 'PUT', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async deletePhotoFromProject(tableName, { photoIds } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/photos/${encodeURIComponent(tableName)}/delete/`;
    const query = new URLSearchParams();
    if (photoIds !== undefined) query.set('photo_ids', photoIds);
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'DELETE', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }
}

File: sdk/v1.0.15/python/mediaviz_sdk/photos.py
from __future__ import annotations
from typing import Any
from urllib.parse import quote, urlencode
import httpx

from .errors import handle_response


class Photos:
    def __init__(self, ctx) -> None:
        self._ctx = ctx

    def add_photo_to_project(
        self,
        photo: Any,
        tableName: Any,
        sourceResolutionX: Any,
        sourceResolutionY: Any,
        dateTaken: Any,
        latitude: Any,
        longitude: Any,
        filePath: Any,
        title: Any,
        clientSideId: Any,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/'
        body = {
            'photo': photo,
            'table_name': tableName,
            'source_resolution_x': sourceResolutionX,
            'source_resolution_y': sourceResolutionY,
            'date_taken': dateTaken,
            'latitude': latitude,
            'longitude': longitude,
            'file_path': filePath,
            'title': title,
            'client_side_id': clientSideId,
        }
        return self._ctx.client.request(path, 'POST', self._ctx.access_token, self._ctx.refresh_token, body).data

    def get_photo_from_project(
        self,
        table_name: str,
        photo_id: int,
        keyword_list_id: int | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/' + quote(str(photo_id), safe='')
        _q: dict[str, str] = {}
        if keyword_list_id is not None:
            _q['keyword_list_id'] = str(keyword_list_id)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_photo_face_details_from_project(self, table_name: str, photo_id: int) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/face_details/' + quote(str(table_name), safe='') + '/' + quote(str(photo_id), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_all_project_photo_ids(
        self,
        table_name: str,
        asc_or_desc: str | None = None,
        last_id: Any | None = None,
        limit: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/'
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if last_id is not None:
            _q['last_id'] = str(last_id)
        if limit is not None:
            _q['limit'] = str(limit)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_top_middle_bottom_project_photo_ids(
        self,
        table_name: str,
        asc_or_desc: str | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos_top/' + quote(str(table_name), safe='')
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_top_middle_bottom_project_photos_by_table_name_sorted_by_date_ranked_keyset_paginated(
        self,
        table_name: str,
        asc_or_desc: str | None = None,
        last_id: Any | None = None,
        limit: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos_ranked/' + quote(str(table_name), safe='') + '/'
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if last_id is not None:
            _q['last_id'] = str(last_id)
        if limit is not None:
            _q['limit'] = str(limit)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_project_month_years_with_photos(self, table_name: str) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photo_month_years/' + quote(str(table_name), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_project_photo_ids_by_month(
        self,
        table_name: str,
        month: int,
        year: int,
        asc_or_desc: str | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/month/' + quote(str(month), safe='') + '/year/' + quote(str(year), safe='')
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_project_photo_ids_by_month_keyset_paginated(
        self,
        table_name: str,
        month: int,
        year: int,
        asc_or_desc: str | None = None,
        last_id: Any | None = None,
        limit: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/month/' + quote(str(month), safe='') + '/year/' + quote(str(year), safe='') + '/'
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if last_id is not None:
            _q['last_id'] = str(last_id)
        if limit is not None:
            _q['limit'] = str(limit)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_top_project_photos_by_table_name_by_month(
        self,
        table_name: str,
        month: int,
        year: int,
        asc_or_desc: str | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/month/' + quote(str(month), safe='') + '/year/' + quote(str(year), safe='') + '/top'
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_top_middle_bottom_project_photos_by_table_name_by_month_sorted_by_date_ranked_keyset_paginated(
        self,
        table_name: str,
        month: int,
        year: int,
        asc_or_desc: str | None = None,
        last_id: Any | None = None,
        limit: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/month/' + quote(str(month), safe='') + '/year/' + quote(str(year), safe='') + '/ranked/'
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if last_id is not None:
            _q['last_id'] = str(last_id)
        if limit is not None:
            _q['limit'] = str(limit)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_project_photo_ids_no_date_taken_keyset_paginated(
        self,
        table_name: str,
        asc_or_desc: str | None = None,
        last_id: Any | None = None,
        limit: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/date_taken/none/'
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if last_id is not None:
            _q['last_id'] = str(last_id)
        if limit is not None:
            _q['limit'] = str(limit)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_top_project_photos_by_table_name_no_date_taken_ranked(
        self,
        table_name: str,
        asc_or_desc: str | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/date_taken/none/ranked'
        _q: dict[str, str] = {}
        if asc_or_desc is not None:
            _q['asc_or_desc'] = str(asc_or_desc)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_project_thumbnail(self, table_name: str) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos_project/' + quote(str(table_name), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def update_photo_in_project(
        self,
        table_name: str | None = None,
        photo_id: int | None = None,
        photo_data: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos_update'
        _q: dict[str, str] = {}
        if table_name is not None:
            _q['table_name'] = str(table_name)
        if photo_id is not None:
            _q['photo_id'] = str(photo_id)
        if photo_data is not None:
            _q['photo_data'] = str(photo_data)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'PUT', self._ctx.access_token, self._ctx.refresh_token).data

    def update_photo_ranking(
        self,
        table_name: str,
        photo_id: int,
        new_category: str,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos_update/' + quote(str(table_name), safe='') + '/id/' + quote(str(photo_id), safe='') + '/rank/' + quote(str(new_category), safe='')
        return self._ctx.client.request(path, 'PUT', self._ctx.access_token, self._ctx.refresh_token).data

    def delete_photo_from_project(
        self,
        table_name: str,
        photo_ids: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/photos/' + quote(str(table_name), safe='') + '/delete/'
        _q: dict[str, str] = {}
        if photo_ids is not None:
            _q['photo_ids'] = str(photo_ids)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'DELETE', self._ctx.access_token, self._ctx.refresh_token).data
